### ISSUE TITLE
question about "trim" option in idft

### ISSUE THREAD

**Issue by m-yoneya** [votca/ctp#85]
Dear all:

In the VOTCA-CTP manual, the default value of "trim" option in the idft is 2.
We'd found this trim=2 setting made large deviations (about two times lager in max.)
of the calculated rates (with g03) from the corresponding values with ADF (Amsterdam Density Function package with corresponding XC function and basis sets etc.).
We'd tried trim=0,1,2,3 and also excluding trim from tasks in idft  and found trim=1 gives
best match with the ADF values (trim=0 made errors).
Could someone kindly gives more info.s about the "trim" option (does it concern only in
idft calculation or something in gaussian calculations?) and guideline to select it.

Regards,

**Comment by 12AngryMen** [comment]
Heyho,

the source code says the following:
"// reduces the number of virtual orbitals to
(trim-1)*number_of_occupied_orbitals".

This is only done for the monomers calculations.

Cheers Jens



2016-03-29 5:25 GMT+02:00 m-yoneya <notifications@github.com>:

**Comment by m-yoneya** [comment]
Thanks for the comment.

As poitend out.
"// reduces the number of virtual orbitals to (trim-1)*number_of_occupied_orbitals".
Then, trim=1 reduces the number of virtual orbitals to 0.
We're wondering whether the trim=1 is valid setting or not.
Excluding trim from tasks in idft also resulted in a large deviation from the ADF value
(then, some trim value would be necessary).

Regards,

**Comment by 12AngryMen** [comment]
To my undestanding,

not specifying trim leads to votca using all virtual orbitals.

Did you calculate it for electrons or holes?

I am a little surprised as well, but fortunately I did not write that part.
:D

Thanks for bringing it up.

Jens

2016-03-29 6:25 GMT+02:00 m-yoneya <notifications@github.com>:

**Comment by m-yoneya** [comment]
Thanks again.

With excluding trim from tasks in idft, the rates were about four times lager in max. compare with ADF value (even worse than the default trim=2).
We're trying hole transport in crystalline pentacene (structure from the CIF file).

Regards,

**Comment by 12AngryMen** [comment]
Hello,

okay that is all still very strange. As I read the code, when trim=1 no virtual levels are taken into account. Okay, can you check the electron couplings? Do they remain the same? Theoretically trim should not matter. Weird.

**Comment by 12AngryMen** [comment]
which version of votca-ctp are you using?

**Comment by 12AngryMen** [comment]
Could you send me the problematic input files for your system to wehnerj@mpip-mainz.mpg.de?

Cheers Jens

**Comment by crionez** [comment]
Hi,

trim option reduces the number of virtual orbitals in the electronic
coupling calculation. If you do not mention it in "tasks", all virtual
orbitals will be used. Otherwise all matrices are resized by trim_factor
* number_of_occupied_levels , i.e. if trim = 2, you will have the same
number of virtual and occupied orbitals. The program does not check if
you have enough virtual orbitals to trim the matrix, this potentially
can lead to problems.

My experience that calculations with trim=2 give very similar results to
the situation when all orbitals are taken into account. You can check if
ADF outputs enough virtual orbitals to begin with.

You can also try "coupling" calculator of ctp_tools, it performs more
checks.

Also add -v option to your ctp_parallel and have a look at matrix sizes
idft outputs - they should be the same for both gausian and idft inputs.

The links to the code and the relevant parts are below.

Hopefully it helps,
Denis

--

//////////////////////////////////////////////////////////////////
https://github.com/votca/ctp/blob/master/src/libctp/jobcalculators/idft.cc
//////////////////////////////////////////////////////////////////

if ( _do_trim ) {
            _orbitalsA.Trim(_trim_factor);
            _orbitalsB.Trim(_trim_factor);
}

//////////////////////////////////////////////////////////////////
https://github.com/votca/ctp/blob/master/src/libctp/orbitals.cc
//////////////////////////////////////////////////////////////////

// reduces the number of virtual orbitals to
// factor*number_of_occupied_orbitals
void Orbitals::Trim( int factor ) {

if ( _has_mo_coefficients ) {
        _mo_coefficients.resize ( factor * _occupied_levels,
_basis_set_size, true);
        _unoccupied_levels = ( factor -1 ) * _occupied_levels;
}

if ( _has_mo_energies ) {
        _mo_energies.resize(  factor * _occupied_levels, true );
        _unoccupied_levels = ( factor - 1) * _occupied_levels;
}
}


On 03/29/2016 05:25 AM, m-yoneya wrote:
> Dear all:
> 
> In the VOTCA-CTP manual, the default value of "trim" option in the idft
> is 2.
> We'd found this trim=2 setting made large deviations (about two times
> lager in max.)
> of the calculated rates (with g03) from the corresponding values with
> ADF (Amsterdam Density Function package with corresponding XC function
> and basis sets etc.).
> We'd tried trim=0,1,2,3 and also excluding trim from tasks in idft and
> found trim=1 gives
> best match with the ADF values (trim=0 made errors).
> Could someone kindly gives more info.s about the "trim" option (does it
> concern only in
> idft calculation or something in gaussian calculations?) and guideline
> to select it.
> 
> Regards,
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/votca/ctp/issues/85>
> 


-- 
Max Planck Institute for Polymer Research
Ackermannweg 10, D-55128 Mainz, Germany
http://www.mpip-mainz.mpg.de/~andrienk/

**Comment by m-yoneya** [comment]
Thanks a lot for many info.s.

The version we're trying is "1.3-dev (27aa2b1) Jan. 15, 2016".
We'll try additional tools and options as suggested.

Best regards,

**Comment by m-yoneya** [comment]
Dear all:

We'd checked the matrix sizes.
As pointed out, trim=2 set the same number of virtual and occupied orbitals.
With trim=1, number of virtual orbitals was zero.
With trim=3, number of virtual orbitals was (3-1) times occupied orbitals.
Then, increasing the number of trim increases the number of virtual orbitals towards
to the full number when trim task was neglected.
If so, trim=1 would be somewhat dangerous setting.
Are these correct?
Anyway, we found the trim setting has most effect to the value of the calculated rates.

Regards,

**Comment by baumeier** [comment]
Yes, this is correct.
I did a test of my own on some pentacene pairs. An ideal stack at 3.7A distance and one at 5A distance. Couplings for different trim options, compared to 0.5*(HOMO-HOMO-1) of the dimer splitting [all values in eV]:

at distance 3.7:
split        0.24966
trim=2   0.26082
trim=3   0.31508
trim=4   0.36343

at distance 5.0:
split        0.03374
trim=2   0.03419
trim=3   0.03447
trim=4   0.03468

It seems that this mirrors your observation. Note that I compared to the splitting because this is an internally consistent reference. What happens in the code is that it determines coupling elements between all monomer levels that survive after trimming, which also implies that it is always taking all occupied levels into account, plus a number of unoccupied ones according to trim. In the process of calculating the effective coupling elements, one has to orthogonalize the monomer orbitals. This step appears to be a bit sensitive on the number of levels, though I am not sure if this is a numerical or conceptional issue. Anyways, it is obviously more significant for strongly coupled systems

What I did now was also to test a “minimal” model that only takes the respective monomer HOMOs and LUMOs into account in the effective coupling:

at distance 3.7:
split        0.24966
minimal 0.24963

at distance 5.0:
split        0.03374
minimal 0.03396

I will add this option to the code and write a note when it’s done.
